pimd/pim_ifchannel.c

```diff
@@ -174,8 +174,11 @@ void pim_ifchannel_delete(struct pim_ifchannel *ch)
 	   ifchannel list is empty before deleting upstream_del
 	   ref count will take care of it.
 	*/
-	pim_upstream_del(pim_ifp->pim, ch->upstream, __PRETTY_FUNCTION__);
-	ch->upstream = NULL;
+
+        if (!(ch->upstream->flags & PIM_UPSTREAM_FLAG_DEL)) {
```

Comment:
1) Why do we need a separate flag here ? Just checking against ch->upstream is NULL or not would be enough. Am I missing something here ? Given a choice, I would like to avoid the flag as it needs separate maintenance & associated code. 
2) I see an issue here. We are setting ch->upstream = NULL. In case this functions gets called again after ch->upstream set to NULL, then this will crash as we are checking ch->upstream->flag. This causes an NULL pointer exception. This may not have showed up in your testing as there may not be any case where this function gets called after setting ch->upstream = NULL. But this should be avoided.